### ISSUE TITLE
$(out) can not access parameters of the output that is a dictionary.

### ISSUE THREAD

**Issue by marko7460** [GoogleCloudPlatform/deploymentmanager-samples#353]
At the moment `$(out)` doesn't support extracting parameter from the output that returns dictionary.

For example take this manifest:
```yaml
outputs:
- finalValue: folders/1111111111111
  name: CORE
  value: $(ref.sato-folders.folders.CORE.name)
resources:
- name: sato-folders
  outputs:
  - finalValue:
      CORE:
        createTime: '2018-11-27T21:58:36.807Z'
        displayName: CORE
        lifecycleState: ACTIVE
        name: folders/111111111
        parent: organizations/1234567890
      EU:
        createTime: '2018-11-27T22:13:47.794Z'
        displayName: EU
        lifecycleState: ACTIVE
        name: folders/22222222222
        parent: organizations/1234567890
      NA:
        createTime: '2018-11-27T21:58:37.313Z'
        displayName: NA
        lifecycleState: ACTIVE
        name: folders/3333333333
        parent: organizations/1234567890
    name: folders
    value:
      CORE:
        createTime: $(ref.CORE.createTime)
        displayName: $(ref.CORE.displayName)
        lifecycleState: $(ref.CORE.lifecycleState)
        name: $(ref.CORE.name)
        parent: $(ref.CORE.parent)
      EU:
        createTime: $(ref.EU.createTime)
        displayName: $(ref.EU.displayName)
        lifecycleState: $(ref.EU.lifecycleState)
        name: $(ref.EU.name)
        parent: $(ref.EU.parent)
      NA:
        createTime: $(ref.NA.createTime)
        displayName: $(ref.NA.displayName)
        lifecycleState: $(ref.NA.lifecycleState)
        name: $(ref.NA.name)
        parent: $(ref.NA.parent)
  properties:
    folders:
    - displayName: NA
      name: NA
      orgId: organizations/1234567890
    - displayName: EU
      name: EU
      orgId: organizations/1234567890
    - displayName: CORE
      name: CORE
      orgId: organizations/1234567890
  resources:
  - name: NA
    type: gcp-types/cloudresourcemanager-v2:folders
  - name: EU
    type: gcp-types/cloudresourcemanager-v2:folders
  - name: CORE
    type: gcp-types/cloudresourcemanager-v2:folders
  type: folder.py
```
If i want to create a project under CORE i would have to do something like this
```yaml
name: host-project
imports:
  - path: ../../../deploymentmanager-samples/community/cloud-foundation/templates/project/project.py
    name: project.py
project: {{env.CFT_BOOTSTRAP_PROJECT_ID}}
resources:
  - name: tools-host
    type: project.py
    properties:
      billingAccountId: "{{env.CFT_BILLING_ACCOUNT_ID}}"
      parent:
        type: folder
        id: $(out.{{env.CFT_BOOTSTRAP_PROJECT_ID}}.folders.sato-folders.folders["CORE"]["name"])
      activateApis:
        - compute.googleapis.com
        - deploymentmanager.googleapis.com
        - pubsub.googleapis.com
        - container.googleapis.com
      serviceAccounts:
        - accountId: cft-demo-sa-0
          displayName: cft demo service account 0
          roles:
            - roles/editor
            - roles/viewer
          networkAccess: true
      usageExportBucket: false
      sharedVPCHost: true
      removeDefaultVPC: true
```
if i place `id: $(out.{{env.CFT_BOOTSTRAP_PROJECT_ID}}.folders.sato-folders.folders)` then value for id will be a map and `cft` will throw an error.

**Comment by ocsig** [comment]
Hi Marko,

I totally understand your point, I've faced the same "limitation" a few weeks ago and yes, DM and CFT references are able to select an output, but unable to process it as a complex structure. We can argue about how painful is this or is the YAML the right place for complex data structure processing or not.

However conceptually CFT templates are not necessary the top level templates what you call directly from the YAMLs. ( There is nothing wrong with it, but for variable injection, name generation via helper function, global configuration it's not necessary the best practice.)

The easiest solution is wrapping the template into an other template:
`

   def generate_config(context):

    # Manipulate context.properties #
    
    resources =[{
      'type': context.properties['cft-type'],
      'name': context.env['name'],
      'properties': context.properties}]
      
    return {'resources': resources}

`

This small template gives you the option to get any input and translate it to the right format. You can also hardcode the 'cft-type' value in your template.

You should ask now, why do you need this pain instead of a simple selector in the yaml ( if that would be implemented)? Let me give you a use case:
You maybe want to combine multiple values, do string manipulation like substring, you may want a standard naming convention, you may want to use global values. All of these in the YAML would result a really overloaded, hard to maintain configuration.

That is the main reason why I encourage you to wrap the CFT templates into your own templates which are adding the logic of your organization.

Here is a bit more complex example how can you load in further python files which may contain (shared/global) properties. Please note, this is just an example, structuring your DM codebase is a much larger engineering effort.
`
import configs/global.py    #contains a python hashmap with globally shared config properties
import configs/department.py
import configs/global.py
import configs/environment.py

def generate_config(context):

    # Manipulate context.properties #
# globals, department, system, environment are hashmaps defined in the included external python files.
context.properties[my_instance][name]= globals[org_name] + '-' + department[department_name] + '-' + system[System_id] + '-' + environment[env_Name]

    
    resources =[{
      'type': context.properties['cft-type'],
      'name': context.env['name'],
      'properties': context.properties}]
      
    return {'resources': resources}
`

If you are interested, here is my tutorial about handling a large codebase with Deployment Manager:
[https://cloud.google.com/deployment-manager/docs/at-scale](https://cloud.google.com/deployment-manager/docs/at-scale)
This is unrelated to the CFT templates, this is a transparent layer of variable injection and helper functions.

If you have any further questions, feel free to reach me out here.

**Comment by ocsig** [comment]
PR #355 got merged into master. This feature adds support to reference cross deployment dictionary outputs.

**Comment by komasoftware** [comment]
@ocsig  I still do not see how this helps getting hold of the folder.id and how you can then manipulate the returned output in the form of "folders/1234567" and convert it to a numeric 1234567 that can be passed as parameter for the project creation.

The below approach does not work :

**environment.py** : create folder en project per environment e.g. dev/test/prod
```
def generate_config(context):
    """Main entry point for Deployment Manager logic"""

    resources = []
    outputs = []

    # environment outputs
    outputs.append({
        "name": "environment_name",
        "value": context.properties["envName"]
    })

    # environment folder
    resources.append({
        "name": "environment_folder",
        "type": "environment_folder.py",
        "properties": context.properties
    })
    outputs.append({
        "name": "environment_folder_id",
        "value": "$(ref.environment_folder.folder_id)"
    })

    # front_end project
    resources.append({
        "name": "front_end",
        "type": "front_end.py",
        "properties": {
            "envName": context.properties["envName"],
            "environment_folder": "$(ref.environment_folder.folder_id)"
        },
        "metadata": {
            "dependsOn": ["environment_folder"]
        }
    })

    return {"resources": resources, "outputs": outputs}

```

**environment folder** : create folder for the environment
```
from helper import config_merger


def generate_config(context):
    """Generate configuration."""

    resources = []
    outputs = []

    # create environment folder
    module = "environment_folder"
    cc = config_merger.ConfigContext(context.properties, module)
    folder_name = context.properties["envName"] + "_folder"
    resources.append({
        "name": folder_name,
        "type": "gcp-types/cloudresourcemanager-v2:folders",
                "properties": {
                    "displayName": cc.configs["Application_Name"] + " " + context.properties["envName"] + " folder",
                    "parent": cc.configs["Parent_Folder_Id"]
                }
    })
    outputs.append({
        "name": "folder_id",
        "value": "$(ref." + folder_name + ".name)"
    })

    return {"resources": resources, "outputs": outputs}
```

**front_end.py** : project to host the frontend application
def GenerateConfig(context):

    resources = []
    outputs = []

    # create front_end project
    module = "front_end"
    cc = config_merger.ConfigContext(context.properties, module)
    project_id = cc.configs["Application_Name"] + \
        "-" + context.properties["envName"] + "-front-end"
    project_name = cc.configs["Application_Name"] + \
        " " + context.properties["envName"] + " front end"
    folder_id = context.properties["environment_folder"]

    resources.append({
        'name': project_id,
        'type': 'cloudresourcemanager.v1.project',
        'properties': {
            'name': project_name,
            'projectId': project_id,
            'parent': {
                'type': 'folder',
                'id': folder_id  // THROWS ERROR, MUST BE REFORMATTED, FINAL VALUE NOT ACCESSBILE
            }
        }
    })

    return {'resources': resources, 'outputs': outputs}

**Comment by ocsig** [comment]
Hi @komasoftware,

Manipulating a referenced value in the same deployment IS NOT possible with the current DM implementation. The interpretation of the referenced value happens only after the python/jinja run.
 #355 an additional wrapper around gcloud which adds cross deployment referencing capabilities and solves this issue, however it is a deprecated python2 implementation, we are developing a Go version, alpha release is here: https://github.com/GoogleCloudPlatform/cloud-foundation-toolkit/tree/cft-dm-dev/cli ( not documented yet.)

That being said, Folders and projects should not be created in the same deployment in a large, well structured system ( with appropriate cross deployment referencing).

**Comment by komasoftware** [comment]
Thx @ocsig for the reply.

- So I would create a deployment that creates the folder. 
- Copy the resulting folder Id into the properties of the template for my project deployment ?

Sticking to vanilla DM (not CFT), this is the only way to achieve a dependency ?

**Comment by ocsig** [comment]
To be really precise:

- When you mean copy the folderID output to the next deployments input, you mean via a small script?  (which is querying and parsing gcloud output.)

Yes, if you don't use the CFT CLI, you need to grab and pass the values between deployments. ( You can still use the CFT templates without the CLI, but it will not solve this issue.)

Note: Folders and Projects should not be created in the same deployment in my opinion. Projects should be separate because they are really complicated resources. I usually advice to have one or more deployments for Folders, one deployment for every project and at least one deployment for the content withitn the project. It needs orchestration between deployments for sure.
